define buffer on heap

cj-matti-niemenmaa opened the pull request:
(PRing this on @james-mathews-cujo's behalf.)

for devices with low stack size we cannot define 64k of message space on the stack

cj-james-mathews committed:
define buffer on heap

for devices with low stack size we cannot define 64k of message space on
the stack

gabriel-cujo commented on src/netlink.c:
```diff
@@ -90,20 +90,27 @@ static int meth_send(lua_State *L) {
+    p_timeout tm = &nl->tm;
     size_t sent = 0;
     int err;
+    struct nlmsgbuf *nlb = malloc(sizeof(struct nlmsgbuf));
```
Have you measured the performance impact of this change? I think you should allocate this buffer only once in the Lua userdata and reuse it. Reusing shouldn't be a problem given that this function is not reentrant and Lua is single-threaded.

cj-matti-niemenmaa replied:
We haven't had the time for a proper performance evaluation of this, no. I feel like it shouldn't be particularly significant, though, since there tends to be a lot of mallocing going on in Lua code anyway and this is in conjunction with a syscall + memcpying so we're doing something relatively expensive regardless.

gabriel-cujo replied:
We are moving towards reducing the memory usage in Lua; reusing tables and so on. In particular, we want to reduce the overhead when receiving messages from the kernel, not add more overhead. If you don't want to bother measuring the change, you could at least perform a change that theoretically won't add overhead. If you still don't want to have this minimal amount of work, you could add an #ifdef that doesn't allocate the buffer in the heap by default and only allocate it there on the necessary platforms.

cj-matti-niemenmaa replied:
Yeah, I do understand your point, and both the userdata and the `#ifdef` alternatives make sense to me.

For the `#ifdef` though we'd have to back and determine which exact platforms need this, which is unfortunately some manual work.

I'll have to get back to this later, once I have some spare time.

(Also I feel like it's worth pointing out that measuring this is not a minimal amount of work, since we don't have well-defined benchmarks for testing pretty much anything and microbenchmarks aren't always indicative of real-world performance, and we'd likely have to do this across several platforms which unfortunately isn't something we've automated to the point of convenience.)

gabriel-cujo replied:
The minimal amount of work is moving these mallocs to the userdata structure, not measuring the performance. This change should be simple, just add another field to the t_netlink struct:
https://github.com/getCUJO/luasocket/blob/a0c55f75e1eaa4c825f40c8611e6111a1e70ccad/src/netlink.h#L32
You don't even need to add a malloc since this object is already on the heap.
```
typedef struct {
    t_socket fd; 
    t_timeout tm;  
    t_pid srcpid; 
    t_type type; 
    struct nlmsgbuf nlmsgbuf; /* this should be very easy */
} t_netlink;
```

cj-matti-niemenmaa replied:
Ah, then I misunderstood, apologies. I didn't realize there was already a struct like that for adding things to, I'm not very familiar with this code. Thanks for the hint, I can probably get that change done and tested soonish.

cj-matti-niemenmaa replied:
Pushed an update with this type of change. I didn't squash it since I felt preserving James's authorship for the original change to be worth leaving it separate.

cj-matti-niemenmaa committed:
netlink: Preallocate message buffer in userdata

Instead of allocating and freeing it on every send/receive call.

While we're touching these lines anyway, do two small cleanups: Cast the
buffer to const char* instead of plain char* when sending, and fix
indentation to consistently use spaces instead of tabs.

gabriel-cujo commented on src/netlink.h:
```diff
@@ -24,11 +26,17 @@ typedef int t_pid;
     t_timeout tm;  
     t_pid srcpid; 
     t_type type; 
+    struct nlmsgbuf *nlb;
```
Why this needs to be a pointer?

cj-matti-niemenmaa replied:
No reason I suppose, we could even inline `nlmsgbuf` entirely at this point. It felt natural to me to keep it separate but I suppose the indirection is a waste since this all gets heap-allocated anyway.

(Sorry for the late reply, first I was sick and then I got swamped with some other stuff.)

cj-carlos-carvalho commented:
@james-mathews-cujo @matti-cujo could we have this PR approved? 

This PR is analogous `cujo-luasocket53` `patches/0002-define-buffer-on-heap.patch` 

Ideally this would be merged after https://github.com/getCUJO/luasocket/pull/7
